src/doc_iter.rs

```diff
@@ -0,0 +1,309 @@
+use crate::document::Document;
+
+impl Document {
+    /// Returns an iterator over all all document nodes including
+    /// comments and fragments.
+    ///
+    /// When encountering a container node (mapping, sequence or fragment),
+    /// the container node is yielded first, then all of its children.
+    pub fn iter(&self) -> DocIter {
+        let v = std::slice::from_ref(self);
+        DocIter {
+            stack: vec![v.iter()],
+        }
+    }
+
+    /// Returns an iterator over all value nodes in the document.
+    /// The iterator yields tuples of (object-path, value-node).
+    pub fn iter_path(&self) -> DocPathIter {
+        let v = std::slice::from_ref(self);
+        DocPathIter {
+            stack: vec![v.iter()],
+            aggregate: Vec::new(),
+            path: Vec::new(),
+        }
+    }
+
+    /// Returns a mutable iterator over all value nodes in the document.
+    /// The iterator yields tuples of (object-path, value-node).
+    pub fn iter_path_mut(&mut self) -> DocPathIterMut {
+        let v = std::slice::from_mut(self);
+        DocPathIterMut {
+            stack: vec![v.iter_mut()],
+            aggregate: Vec::new(),
+            path: Vec::new(),
+        }
+    }
+}
+
+impl<'a> IntoIterator for &'a Document {
+    type Item = &'a Document;
+    type IntoIter = DocIter<'a>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        self.iter()
+    }
+}
+
+pub struct DocIter<'a> {
+    stack: Vec<std::slice::Iter<'a, Document>>,
+}
+
+impl<'a> Iterator for DocIter<'a> {
+    type Item = &'a Document;
+    fn next(&mut self) -> Option<Self::Item> {
+        while let Some(top) = self.stack.last_mut() {
+            if let Some(val) = top.next() {
+                match val {
+                    Document::Mapping(v) => {
+                        self.stack.push(v.iter());
+                        return Some(val);
+                    }
+                    Document::Sequence(v) => {
+                        self.stack.push(v.iter());
+                        return Some(val);
+                    }
+                    Document::Compact(v) => {
+                        self.stack.push(std::slice::from_ref(&**v).iter());
+                        return Some(val);
+                    }
+                    Document::Fragment(v) => {
+                        self.stack.push(v.iter());
+                        return Some(val);
+                    }
+                    _ => return Some(val),
+                }
```

Comment:
Every branch of this `match` ends with `return Some(val)`, so you move that after the `match`:

```rust
        match val {
            Document::Mapping(v) => {
                self.stack.push(v.iter());
            }
            Document::Sequence(v) => {
                self.stack.push(v.iter());
            }
            Document::Compact(v) => {
                self.stack.push(std::slice::from_ref(&**v).iter());
            }
            Document::Fragment(v) => {
                self.stack.push(v.iter());
            }
            _ => {},
        }
        return Some(val);
```

If you apply my above suggestion to remove the nesting, `return Some(val);` can just be `Some(val)`.